File: src/main/scala/com/github/ahnfelt/topshell/language/Tokenizer.scala
package com.github.ahnfelt.topshell.language

import com.github.ahnfelt.topshell.language.Syntax.Location

object Tokenizer {

    private val patterns = """
        comment:    [/][/][^\r\n]*
        number:     [0-9]+(?:[.][0-9]+(?:[eE][+-]?[0-9]+)?)?\b
        string:     ["](?:[^"\\\r\n]+|[\\][\\"trn])*["]
        lower:      [a-z][a-zA-Z0-9]*\b
        upper:      [A-Z][a-zA-Z0-9]*\b
        wildcard:   [_]\b
        operator:   [!@#$%&/=?+|^~*<>.:-]+
        separator:  [,;]|[.][.]?|[:][:]?
        bracket:    [(){}\[\]]
        space:      [ ]+
        newline:    [\r]?[\n]
        unknown:    [^\n ]+
    """

    private val (keys, parts) = patterns.split("\n").filter(_.trim.nonEmpty).map(_.split(":", 2).map(_.trim)).map {
        case Array(k, v) => k -> v
        case _ => throw new RuntimeException("The patterns variable is not correctly formatted.")
    }.unzip

    private val pattern = parts.map("(" + _ + ")").mkString("|").r

    private val groups = keys.zipWithIndex.map { case (k, i) => (k, i + 1) }.toMap

    case class Token(at : Location, kind : String, raw : String)

    case class ParseException(at : Location, message : String) extends RuntimeException(message + " " + at)

    def tokenize(file : String, code : String) : Array[Token] = {
        val tokens = {
            var line = 1
            var lineStart = 0
            var wildcard = 0
            pattern.findAllMatchIn(code).map { m =>
                def location = Location(file, line, 1 + m.start - lineStart)
                val token = if(m.group(groups("space")) != null) {
                    List.empty
                } else if(m.group(groups("newline")) != null) {
                    line += 1
                    lineStart = m.end
                    List.empty
                } else if(m.group(groups("comment")) != null) {
                    List.empty
                } else if(m.group(groups("lower")) != null) {
                    List(Token(location, "lower", m.group(groups("lower"))))
                } else if(m.group(groups("upper")) != null) {
                    List(Token(location, "upper", m.group(groups("upper"))))
                } else if(m.group(groups("operator")) != null) {
                    val raw = m.group(groups("operator"))
                    val kind =
                        if(List("=", "<-", "->", ".", "..", ":", "::", "?").contains(raw)) "separator" else "operator"
                    List(Token(location, kind, raw))
                } else if(m.group(groups("wildcard")) != null) {
                    wildcard += 1
                    List(Token(location, "definition", "_" + wildcard))
                } else {
                    val iterator = for(k <- keys) yield Option(m.group(groups(k))).map(g => Token(location, k, g))
                    List(iterator.collectFirst { case Some(t) => t }.get)
                }
                val all = m.group(0)
                if(m.start == lineStart && all != "}" && all != "]" && all != ")" && all.trim.nonEmpty) {
                    Token(location, "top", "top level symbol") :: token
                } else token
            }
        }.flatten.toArray
        tokens.zip(tokens.drop(1) :+ Token(null, "internal", "internal")).map { case (token, next) =>
            if(token.kind == "lower" && Seq("=", "<-", "->", "::").contains(next.raw)) {
                token.copy(kind = "definition")
            } else if(token.kind == "upper" && Seq("@").contains(next.raw)) {
                token.copy(kind = "definition")
            } else {
                token
            }
        }
    }

}

File: tokenizer.js
self.tsh.Token = function(
    file, code, kind, startLine, startLineOffset, startOffset, stopLine, stopLineOffset, stopOffset
) {
    this.file = file;
    this.code = code;
    this.kind = kind;
    this.startLine = startLine;
    this.startLineOffset = startLineOffset;
    this.startOffset = startOffset;
    this.stopLine = stopLine;
    this.stopLineOffset = stopLineOffset;
    this.stopOffset = stopOffset;
};

self.tsh.Token.end = (file, code) => new self.tsh.Token(file, code, "end", 0, 0, 0, 0, 0, 0);

self.tsh.tokenize = (file, code) => {
    var tokens = [];
    var line = 1;
    var lineOffset = 0;

    var operatorCharactersString = "!@#$%&/=?+|^~*<>.:-,;";
    var operatorCharacters = {};
    for(var j = 0; j < operatorCharactersString.length; j++) operatorCharacters[operatorCharactersString[j]] = true;

    function token(kind, startOffset, stopOffset) {
        return new self.tsh.Token(
            file, code, kind, startLine, startLineOffset, startOffset, line, lineOffset, stopOffset
        );
    }

    for(var i = 0; i < code.length;) {

        while(code[i] === ' ' || code[i] === '\t' || code[i] === '\r') i++;

        var start = i;
        var startLine = line;
        var startLineOffset = lineOffset;

        if(i === lineOffset && code[i] !== "\n" &&
            code[i] !== "}" && code[i] !== "]" && code[i] !== ")" && code[i] !== "|" && code[i] !== "/"
        ) {
            tokens.push(token("top", i, i));
        }

        if(code[i] === '\n') {

            i += 1;
            line += 1;
            lineOffset = i;

        } else if(code[i] === '/' && code[i + 1] === '/') {

            i += 2;
            while(code[i] !== '\n' && i < code.length) i += 1;

        } else if(code[i] === '/' && code[i + 1] === '*') {

            i += 2;
            while(code[i] !== '*' || code[i + 1] !== '/') {
                if(i >= code.length) {
                    throw 'Expected end of comment started on line ' + startLine + ', got end of file.';
                }
                if(code[i] === '\n') {
                    line += 1;
                    lineOffset = i + 1;
                }
                i += 1;
            }
            i += 2;

        } else if(code[i] === '"') {

            i += 1;
            while(code[i] !== '"') {
                if(code[i] === '\n') {
                    throw 'Unexpected end of line in string started on line ' + startLine + '.';
                }
                if(i >= code.length) {
                    throw 'Expected end of string started on line ' + startLine + ', got end of file.';
                }
                if(code[i] === '\\' && code[i + 1] !== '\n') i += 1;
                i += 1;
            }
            i += 1;
            tokens.push(token("string", start, i));

        } else if((code[i] >= 'a' && code[i] <= 'z') || (code[i] >= 'A' && code[i] <= 'Z')) {

            var kind = code[i] >= 'a' ? "lower" : "upper";
            i += 1;
            while(
                (code[i] >= 'a' && code[i] <= 'z') ||
                (code[i] >= 'A' && code[i] <= 'Z') ||
                (code[i] >= '0' && code[i] <= '9')
                ) i += 1;
            tokens.push(token(kind, start, i));

        } else if(code[i] >= '0' && code[i] <= '9') {

            var dot = false;
            var exponent = false;
            while(code[i] >= '0' && code[i] <= '9') {
                i += 1;
                if((code[i] === 'e' || code[i] === 'E') && !exponent) {
                    i += 1;
                    dot = true;
                    exponent = true;
                    if(code[i] === '+' || code[i] === '-') i += 1;
                }
                if(code[i] === '.' && !dot && !exponent) {
                    i += 1;
                    dot = true;
                }
            }
            tokens.push(token(dot || exponent ? "float" : "int", start, i));

        } else if(code[i] === '_') {

            i += 1;
            tokens.push(token("definition", start, i));

        } else if(operatorCharacters[code[i]] === true) {

            i += 1;
            while(operatorCharacters[code[i]] === true) i += 1;
            var op = code.slice(start, i);
            var operatorKind =
                op === "=" || op === "<-" || op === "->" ||
                op === "," || op === ";" ||
                op === "." || op === ".?" ||
                op === ".." || op === ":" || op === "::" || op === "=>" || op === "?" ?
                    "separator" : "operator";
            tokens.push(token(operatorKind, start, i));

        } else if(
            code[i] === '(' || code[i] === ')' ||
            code[i] === '[' || code[i] === ']' ||
            code[i] === '{' || code[i] === '}'
        ) {

            i += 1;
            tokens.push(token("bracket", start, i));

        } else if(i < code.length) {

            throw 'Unexpected character: ' + code[i];

        }

    }

    for(var k = 0; k + 1 < tokens.length; k++) {
        var previous = k === 0 ? "" : code.slice(tokens[k - 1].startOffset, tokens[k - 1].stopOffset);
        var next = code.slice(tokens[k + 1].startOffset, tokens[k + 1].stopOffset);
        var isField = previous === "{" || previous === "." || previous === ".?" || previous === ",";
        if(tokens[k].kind === "lower" && (next === "=" || next === "<-" || next === "->")) {
            tokens[k].kind = "definition";
        } else if(tokens[k].kind === "lower" && !isField && next === ":") {
            tokens[k].kind = "definition"
        } else if(tokens[k].kind === "upper" && next === "@") {
            tokens[k].kind = "definition"
        }
    }

    return tokens;

};
